gui/src/components/logs/parseSpans.ts

```diff
@@ -1,44 +1,239 @@
-export type Span = Plain | Link;
-
-export interface SpanBase {
-  type: string;
+export interface Span {
+  href?: string;
   text: string;
+  foreground?: string;
+  background?: string;
+  style?: Style;
 }
 
-export interface Plain extends SpanBase {
-  type: 'plain';
-}
+export type Style =
+  | 'bold'
+  | 'faint'
+  | 'italic'
+  | 'underline'
+  | 'blink'
+  | 'invert'
+  | 'strike';
 
-export interface Link extends SpanBase {
-  type: 'link';
-  href: string;
-}
+const isLinkOpenBoundary = (c: string) => {
+  switch (c) {
+    case '':
+    case ' ':
+    case '\n':
+    case '<':
+      return true;
+    default:
+      return false;
+  }
+};
+
+const isLinkCloseBoundary = (c: string) => {
+  switch (c) {
+    case '':
+    case ' ':
+    case '\n':
+    case '<':
+      return true;
+    default:
+      return false;
+  }
+};
+
+const standardColors: string[] = [
+  'rgb(0, 0, 0)', // Black
+  'rgb(170, 0, 0)', // Red
+  'rgb(0, 170, 0)', // Green
+  'rgb(170, 85, 0)', // Yellow
+  'rgb(0, 0, 170)', // Blue
+  'rgb(170, 0, 170)', // Magenta
+  'rgb(0, 170, 170)', // Cyan
+  'rgb(170, 170, 170)', // White
+];
+
+const brightColors: string[] = [
+  'rgb(85, 85, 85)', // Gray
+  'rgb(255, 85, 85)', // Red
+  'rgb(85, 255, 85)', // Green
+  'rgb(255, 255, 85)', // Yellow
+  'rgb(85, 85, 255)', // Blue
+  'rgb(255, 85, 255)', // Magenta
+  'rgb(85, 255, 255)', // Cyan
+  'rgb(255, 255, 255)', // White
+];
+
+const lookupColor256 = (i: number): string => {
+  if (i < 0) {
+    return '#000000';
+  }
+  if (i < 8) {
+    return standardColors[i];
+  }
+  if (i < 16) {
+    return brightColors[i - 8];
+  }
+  if (i < 232) {
+    // Color cube of size 6.
+    // TODO: Does this produce the correct colors?
+    const x = i - 16;
+    const r = ((x >> 4) % 6) * 51;
+    const g = ((x >> 2) % 6) * 51;
+    const b = ((x >> 0) % 6) * 51;
+    return `rgb(${r},${g},${b})`;
+  }
+  if (i < 256) {
+    // Grayscale.
+    const x = ((i - 232) / 24) * 256;
+    return `rgb(${x}, ${x}, ${x})`;
+  }
+  return '000000';
+};
+
+const lookupColor24bit = (codes: number[]): string => {
+  const r = 200;
+  const g = 200;
+  const b = 200;
+  return `rgb(${r}, ${g}, ${b})`;
+};
+
+const lookupColor = (codes: number[]) => {
+  switch (codes[1]) {
+    case 5:
+      return lookupColor256(codes[2]);
+    case 2:
+      return lookupColor24bit(codes);
+    default:
+      return '000000';
+  }
+};
 
 export const parseSpans = (input: string): Span[] => {
```

Comment:
Not convinced by the name since it sounds like the function is parsing spans. I'd prefer `parseLogLine` or `logLineToSpans` or something.

Reply:
Huh, I guess I never thought about the distinction between `parse${source}` and `parse${grammarProduction}` in naming conventions, but upon reflection, I've definitely seen both of these. The latter is much more common IME, so going to leave this alone.